reference/shaders-msl/vert/transform-feedback-decorations.xfb-triangle-list.vert

```diff
@@ -0,0 +1,70 @@
+#pragma clang diagnostic ignored "-Wmissing-prototypes"
+
+#include <metal_stdlib>
+#include <simd/simd.h>
+
+using namespace metal;
+
+struct VertOut
+{
+    float4 vBar;
+};
+
+struct spvXfbBuffer1
+{
+    char _m0_pad[4];
+    packed_float4 gl_Position;
+};
+
+struct spvXfbBuffer2
+{
+    char _m0_pad[16];
+    float4 vFoo;
+};
+
+struct spvXfbBuffer3
+{
+    float4 vBar;
+};
+
+static inline __attribute__((always_inline))
+void foo(thread float4& vFoo)
+{
+    vFoo = float4(3.0);
+}
+
+kernel void main0(uint3 gl_GlobalInvocationID [[thread_position_in_grid]], uint3 spvStageInputSize [[grid_size]], device atomic_uint* spvXfbCounter1 [[buffer(17)]], device spvXfbBuffer1* spvXfb1 [[buffer(13)]], device atomic_uint* spvXfbCounter2 [[buffer(18)]], device spvXfbBuffer2* spvXfb2 [[buffer(14)]], device atomic_uint* spvXfbCounter3 [[buffer(19)]], device spvXfbBuffer3* spvXfb3 [[buffer(15)]])
+{
+    spvXfbBuffer1 spvXfbOutput1 = {};
+    spvXfbBuffer2 spvXfbOutput2 = {};
+    spvXfbBuffer3 spvXfbOutput3 = {};
+    VertOut _25 = {};
+    if (any(gl_GlobalInvocationID >= spvStageInputSize))
+        return;
+    spvXfbOutput1.gl_Position = float4(1.0);
+    foo(spvXfbOutput2.vFoo);
+    _25.vBar = float4(5.0);
+    spvXfbOutput3.vBar = _25.vBar;
+    uint spvXfbIndex = gl_GlobalInvocationID.y * (spvStageInputSize.x - spvStageInputSize.x % 3u) + gl_GlobalInvocationID.x;
+    uint spvInitOffset1 = atomic_load_explicit(spvXfbCounter1, memory_order_relaxed);
+    spvXfb1 = reinterpret_cast<device spvXfbBuffer1*>(reinterpret_cast<device char*>(spvXfb1) + spvInitOffset1);
+    if ((gl_GlobalInvocationID.x % 3u == 2) || gl_GlobalInvocationID.x + 2 < spvStageInputSize.x)
+        spvXfb1[spvXfbIndex] = spvXfbOutput1;
+    uint spvInitOffset2 = atomic_load_explicit(spvXfbCounter2, memory_order_relaxed);
+    spvXfb2 = reinterpret_cast<device spvXfbBuffer2*>(reinterpret_cast<device char*>(spvXfb2) + spvInitOffset2);
+    if ((gl_GlobalInvocationID.x % 3u == 2) || gl_GlobalInvocationID.x + 2 < spvStageInputSize.x)
+        spvXfb2[spvXfbIndex] = spvXfbOutput2;
+    uint spvInitOffset3 = atomic_load_explicit(spvXfbCounter3, memory_order_relaxed);
+    spvXfb3 = reinterpret_cast<device spvXfbBuffer3*>(reinterpret_cast<device char*>(spvXfb3) + spvInitOffset3);
+    if ((gl_GlobalInvocationID.x % 3u == 2) || gl_GlobalInvocationID.x + 2 < spvStageInputSize.x)
+        spvXfb3[spvXfbIndex] = spvXfbOutput3;
+    threadgroup_barrier(mem_flags::mem_device);
```

Comment:
To make sure the atomic load of the counter **happens-before** the atomic update, particularly since Metal doesn't support acquire/release semantics for atomics.

Reply:
Only one thread in the entire dispatch is allowed to write the counter.

...But if the first workgroup completes before the others, they may wind up loading the counter *after* the first thread writes to it. Hmm, this may be tougher than I thought...

Reply:
The only reasonable solution is to manage the counter from the outside and only pass the offset (read-only) into the shader.